### PR TITLE
Implement ScriptSimpleFunctionObject for enhance function call perfor…

### PR DESCRIPTION
**Pull request by ksh8281**
…mance

many general function can use ScriptSimpleFunctionObject.
It has simple call operation

Signed-off-by: Seonghyun Kim <sh8281.kim@samsung.com>

### CODE REVIEW

File: src/runtime/PointerValue.h [hunk 1]
```diff
@@ -163,6 +156,11 @@ class PointerValue : public gc {
         return hasTag(g_doubleInEncodedValueTag);
     }
 
+    inline bool hasArrayObjectTag() const
+    {
+        return hasTag(g_arrayObjectTag);
```
Comment (clover2123) [inline src/runtime/PointerValue.h:161]: Can we replace `hasArrayObjectTag` with existing `isArrayObject` method?
ArrayPrototypeObject is also ArrayObject.
```c++
    inline bool isArrayObject() const
    {
        return hasTag(g_arrayObjectTag) || hasTag(g_arrayPrototypeObjectTag);
    }
```

File: src/runtime/ScriptFunctionObject.cpp [hunk 2]
```diff
@@ -77,6 +78,47 @@ NEVER_INLINE void ScriptFunctionObject::generateByteCodeBlock(ExecutionState& st
...
+        size_t stackStorageSize = cb->totalStackAllocatedVariableSize();
+        size_t literalStorageSize = byteCb->m_numeralLiteralData.size();
+        size_t registerFileSize = registerSize + stackStorageSize + literalStorageSize;
+
+        bool isStrict = cb->isStrict();
+        bool shouldClearStack = byteCb->m_shouldClearStack;
+
+#define DEFINE_SCRIPTSIMPLEFUNCTION_BRANCH(opt1, opt2)                   \
+    if (registerFileSize <= 4) {                                         \
+        writeTag(ScriptSimpleFunctionObject<opt1, opt2, 4>().getTag());  \
```
Comment (clover2123) [inline src/runtime/ScriptFunctionObject.cpp:97]: I'm just wondering that it is really safe to rewrite the vtable address in the middle.
I'm not sure but would it incur any undesired side effect?

File: src/runtime/PointerValue.h (hunk 1 above)
Comment (ksh8281) [reply src/runtime/PointerValue.h:161]: This is fast path for interpreter. I just want to test with 'g_arrayObjectTag'

File: src/runtime/ScriptFunctionObject.cpp (hunk 2 above)
Comment (ksh8281) [reply src/runtime/ScriptFunctionObject.cpp:97]: IMO If sizeof ScriptSimpleFunctionObject and ScriptFunctionObject are same, rewrite is safe.
The only different content of vtable is ::call method address

Comment (clover2123) [reply src/runtime/ScriptFunctionObject.cpp:97]: Ok I see!

**Review by clover2123 — APPROVED** [review]

**Review by bbrto21 — APPROVED** [review]

**Pull request merged**